flink-ml-core/src/main/java/org/apache/flink/ml/common/datastream/DataStreamUtils.java

```diff
@@ -0,0 +1,243 @@
+     * one double array in each partition. The result data stream has the same parallelism as the
+     * input, where each partition contains one double array that sums all of the double arrays in
+     * the input data stream.
+     *
+     * <p>Note that we throw exception when one of the following two cases happen:
+     * <li>There exists one partition that contains more than one double array.
+     * <li>The length of the double array is not consistent among all partitions.
+     *
+     * @param input The input data stream.
+     * @return The result data stream.
+     */
+    public static DataStream<double[]> allReduceSum(DataStream<double[]> input) {
+        return AllReduceImpl.allReduceSum(input);
+    }
+
+    /**
+     * Collects distinct values in a bounded data stream. The parallelism of the output stream is 1.
+     *
+     * @param <T> The class type of the input data stream.
+     * @param input The bounded input data stream.
+     * @return The result data stream that contains all the distinct values.
+     */
+    public static <T> DataStream<T> distinct(DataStream<T> input) {
+        return input.transform(
+                        "distinctInEachPartition",
+                        input.getType(),
+                        new DistinctPartitionOperator<>())
+                .setParallelism(input.getParallelism())
+                .transform(
+                        "distinctInFinalPartition",
+                        input.getType(),
+                        new DistinctPartitionOperator<>())
+                .setParallelism(1);
+    }
+
+    /**
+     * Applies a {@link MapPartitionFunction} on a bounded data stream.
+     *
+     * @param input The input data stream.
+     * @param func The user defined mapPartition function.
+     * @param <IN> The class type of the input element.
+     * @param <OUT> The class type of output element.
+     * @return The result data stream.
+     */
+    public static <IN, OUT> DataStream<OUT> mapPartition(
+            DataStream<IN> input, MapPartitionFunction<IN, OUT> func) {
+        TypeInformation<OUT> resultType =
+                TypeExtractor.getMapPartitionReturnTypes(func, input.getType(), null, true);
+        return input.transform("mapPartition", resultType, new MapPartitionOperator<>(func))
+                .setParallelism(input.getParallelism());
+    }
+
+    /**
+     * Sorts the elements in each partition of the input bounded data stream.
+     *
+     * @param input The input data stream.
+     * @param comparator The comparator used to sort the elements.
+     * @param <IN> The class type of input element.
+     * @return The sorted data stream.
+     */
+    public static <IN> DataStream<IN> sortPartition(
+            DataStream<IN> input, TypeComparator<IN> comparator) {
+        return input.transform(
+                        "sortPartition", input.getType(), new SortPartitionOperator<>(comparator))
+                .setParallelism(input.getParallelism());
```

Comment:
I have removed distinct() and sortPartition() for now.

BTW, `setParallelism()` is needed as we need to make sure that we do not inucr a all-to-all connection, to protect the semantic of `sortPartition` and also for performance.

Reply:
Once we split the global partition (where all elements are sorted) to multiple partitions, it is not clear to me why we even need to have this global partition.

Since this method is removed, maybe we can leave the discussion to future PR.

Reply:
The same problem still exists in `DataStreamUtils#mapPartition`.

When we create a new stream operator in Flink, the parallelism of the new operator is set as `env.getParallelism()`. If the parallelism of the input operator is different from `env.getParallelism()` and we do not call `setParallelism(input.getParallelism)` for the newly created stream operator, we may encounter all-to-all connections between the input operator and newly created operator, which is inconsistent with logic of `mapPartition`.

Reply:
It is useful to know that `datastream.transform(...)` will change the output parallelism to `env. getParallelism(...)`. Thanks for the explanation.

The question regarding why "we need to split the global partition (where all elements are sorted) to multiple partitions" still exists. We can discuss it in the future.